### PR TITLE
Clear DISABLED event on fighters after the carriers repair them.

### PR DESCRIPTION
**Pull request by UnorderedSigh**
**Bugfix:** Second half of this bug fix: #7848 and solves #7662 a different way than #7845

This is one of three PRs that, together, should fix issue #7862

**Important: There are two competing solutions. Please discuss them here: https://github.com/endless-sky/endless-sky/issues/7862**

## Fix Details
Adds a new REPAIRED_IN_BAY ShipEvent sent when a ship in a bay is repaired, so the npc ship's DISABLED flag will be cleared.

ALSO, causes the events from the first frame to be processed by missions. Previously, this wasn't the case. I had to fix that so the ships that are repaired on the first frame get recorded. It took a while to track down that bug...

## Testing Done
Tried the save file a few times. The DISABLED event is reliably cleared. I can't think of any way this could break something else.

## Save File
This save file can be used to verify the bugfix. The bug will occur when using f5e4503a3 but not in this branch.

[Friction Diction~3026-01-21  should work but does not.txt](https://github.com/endless-sky/endless-sky/files/10204247/Friction.Diction.3026-01-21.should.work.but.does.not.txt)

If #7845 is merged, then this save file cannot be used to verify the fix, since that PR makes it so the fighter having the DISABLED trait won't stop its npc fleet from succeeding the "accompany" condition.

### CODE REVIEW

**Review by quyykk — COMMENTED** [review]
Is there a reason why you don't want to use ASSIT?

We have already have a place where a BOARD/ASSIST is generated when a ship boards another one, so why not reuse that part of the code?

**Comment by UnorderedSigh** [comment]
Yes. ASSIST is used to trigger missions, so npc assist is blocked in some places.

**Comment by UnorderedSigh** [comment]
At long last, I figured out why it did not work. The MainPanel tells the Engine to run one step before the PlanetPanel closes, but it ignores any engine.Events() from that step. All fighters are repaired in that first step, so the REPAIRED_IN_BAY event never gets to the MainPanel::StepEvents.

File: source/Engine.h [hunk 1]
```diff
@@ -251,6 +251,9 @@ class Engine {
 	double load = 0.;
 	int loadCount = 0;
 	double loadSum = 0.;
+
+	// For keeping track of which ships were repaired in their carrier's bays
+	std::vector<std::pair<std::shared_ptr<Ship>,std::shared_ptr<Ship>>> repairedInBay;
```
Comment (RisingLeaf) [inline source/Engine.h:256]: This should be moved, I think it would fit to the newShips and that stuff

File: source/NPC.cpp [hunk 2]
```diff
@@ -69,6 +69,8 @@ void NPC::Load(const DataNode &node)
 			succeedIf |= ShipEvent::CAPTURE;
 		else if(node.Token(i) == "provoke")
 			succeedIf |= ShipEvent::PROVOKE;
+		else if(node.Token(i) == "repaired_in_bay")
```
Comment (RisingLeaf) [inline source/NPC.cpp:72]:
```suggestion
		else if(node.Token(i) == "repaired in bay")
```
underscores are not really our style?

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
looks good overall, some smaller things

File: source/Engine.h (hunk 1 above)
Comment (UnorderedSigh) [reply source/Engine.h]: I moved it. How does it look now?

**Comment by vitalchip** [comment]
Looks like you've put a lot of work into this, so kudos.

However, it doesn't fix the bug, it massages the symptoms. That's fine though if we want the change in behaviour.

The mission doesn't complete due to disabled drones, that;s correct. It shouldn't. What should happen next is that you depart and either the player or the NPC fleet or both board and repair the drones and this  will allow the mission to complete on the next landing.

This can't happen though because on departure the disabled drones which should still be in space awaiting assist get loaded onto the carriers (the bug) and now they'd need to launch, be disabled again and *then assisted* before they'd allow the mission to succeed and that's vanishingly unlikely.

This PR will allow the mission to complete by effectively assisting the drone inside the carrier but what it also does is create a further exploit for the player, it now doesn't matter how many carried ships get disabled because they'll all be brought back to health *without* having to go through the normal gameplay of assisting each one.

Stopping the disabled drones being loaded onto the carrier will leave them in space and the NPC fleet will repair them, as soon as they're active they'll dock and be recharged as usual, depending on the carriers ability. 

So, do we want to make the symtoms go away and give the player another exploit or do we want to fix a bug?

I've no skin in the game, I'm on older code and not updating anymore but I think its important to at least be aware of what we're doing and what we're not doing and what the future ramifications might be.

**Comment by UnorderedSigh** [comment]
@vitalchip

You're slightly off on a few things.

1. The DISABLED flag is only relevant to npcs, not player ships or randomly spawned ships.
2. Fixing it as you describe is already the plan, but the problem is that there are numerous corner cases, some we probably haven't even found yet. This is a safeguard to handle cases that haven't fixed yet. This was discussed in #7848
3. The player already has to repair fighters before docking them, so there is no game exploit.
4. This is the only way to fix broken saves where a disabled fighter is already in a carrier.

**Comment by UnorderedSigh** [comment]
@vitalchip For your convenience, I've copied the relevant comment from #7848

> Actually, the ideal solution is to implement both fixes:
> 
> 1. A carrier should board its disable fighters.
> 2. If the disabled fighter ends up docked somehow, it should be automatically repaired.
> 
> This way, you get the pilot confusion on the npc ship when it repairs the disabled fighters (same disadvantage the player flagship has) and you fix the various other situations this could happen.

**Comment by vitalchip** [comment]
Well, you've not read or not understood what I wrote.Like I say I have no issue one way or the other since this code will not be in my code but it does leave a player exploit. Players should have to either assist in the repair of the disabled NPC ships or at least wait for the rest of the fleet to do so. Now they'll just be able to land without caring about any NPC carried ship.

Carrier already repair ship which are placed into them in a disabled state, so the other PR does nothing but give them a free head start.

I know what the disabled flag is and what it applies to, I know that's what's stopping the mission being completed. I also know that it is and is supposed to be undone by assisting, which as I said will be carried out by the rest of the fleet if the disabled carried ships are prevented from being placed on the carriers (which IS the bug).

4. Yes, true but we can easily help anyone stuck to quickly modify their save and it won't happen again once fixed.

Take or leave it. I'm not about to drag a horse to water.

**Comment by UnorderedSigh** [comment]
> Players should have to either assist in the repair of the disabled NPC ships

Once the disabled ship is already in another government's carrier, the player cannot repair it.

**Comment by UnorderedSigh** [comment]
> Yes, true but we can easily help anyone stuck to quickly modify their save and it won't happen again once fixed

Players should not have to edit their saves to fix a bug the came created, when we can easily fix it in code.

Also, it is not easy for everyone to do that.

**Comment by UnorderedSigh** [comment]
> I'm not about to drag a horse to water.

Rudeness like this is uncalled-for. I'm happy to receive reviews and comments, and you obviously have valuable experience to lend. However, you should maintain some decorum, rather than making people feel "less than you" or unwelcome.

**Comment by UnorderedSigh** [comment]
Also, I think the carrier *is* already boarding the fighter, because you can visually see it doing so. I suspect the boarding logic is wrong, and the carrier is docking the fighter instead of boarding it. If I could find the boarding code, I would submit a fix for that, but the code thus far eludes me.

**Comment by UnorderedSigh** [comment]
> It gives you exactly the same amount of repair you would have if the carrier had boarded the ship.

I ensured it gives you the same amount of repair by moving the repair-by-assistance logic to the new Ship::AssistedRepair()

```
// Another ship is helping this disabled ship repair its damage.
void Ship::AssistedRepair()
{
	hull = min(max(hull, MinimumHull() * 1.5), attributes.Get("hull"));
	isDisabled = false;
}
```

**Comment by UnorderedSigh** [comment]
From reading Ship::Board, and viewing some Korath carriers boarding fighters at Misam, I'm pretty sure the carrier in-space boarding *is* working correctly. There is no special case for a carrier boarding a fighter that is disabled and carryable, and the fighter won't board if it is disabled.

That means this probably happened by having the ship automatically be shoved into its parent when the player landed on Stormhold while a fighter was disabled. There are many places where the code shoves fighters into carriers without checking if they're disabled.

**Comment by UnorderedSigh** [comment]
Alternative fix is here: #7866

**Comment by UnorderedSigh** [comment]
Closing due to merge of #7866

**Pull request closed without merging**